### ISSUE TITLE
Consider removing built-in Promise support in favor of `until()` directive.

### ISSUE THREAD

**Issue by justinfagnani** [Polymer/lit-html#550]
Currently we don't support Promises in attributes (#272, #544). I've been doing some work to support this, but it starts getting pretty complex as we try to answer things like:

- What is the value of a fully-controlled attribute before Promise resolution? Empty string or not present?
- What is the value of an interpolated expression before Promise resolution?
- What about boolean attributes? If the default value is not present, that's the same as defaulting to `false`. Is that correct?
- Do Promises also auto-resolve in property bindings? If so, how would we pass Promises as properties to components that actually _accept_ a promise?

While Promises in Node parts generally have none of these problems, support there sets the expectation that Promises just work in attributes as well. Promises support in NodePart also currently has some bugs due to race conditions. We have outstanding code to fix this, but it adds complexity to core.

So, one solution that limits complexity and keeps support consistent is to remove built-in Promise support from NodePart and require the `until()` directive to use Promises. This means that templates have to be explicitly aware of async values, which is also probably a good thing.

`until()` itself would get more complicated as it can't rely on built-in promise support, and has to track Promise identity and resolved state, but that complexity is pay-for-play.

This is obviously a breaking change, so deciding on this blocks a 1.0 release. I'm curious if anyone in the community has feedback here. Upgrading to a new version shouldn't be very difficult if you know where Promises are used in templates.

**Comment by 43081j** [comment]
I think it is better to require use of `until` so people get into the habit of always providing a default value (even if it is empty string).

I don't think it is worth the complexity and flaky-ness of trying to guess what a default value should be. We would end up deciding that an unresolved promise renders as X but its almost guaranteed _many_ people would rather it be Y. 

It is better to be explicit than not.

**Comment by daKmoR** [comment]
One use case where we use promises a LOT is for translations.
```js
const msg = () => { return Promise(..) }

render(html`
  <my-el label=${msg('my-el:label')}> // that does not work currently
     <h2>${msg('my-el:heading')}></h2> // works fine
  <my-el>
`, target);
```

See "working" example:
https://codepen.io/daKmoR/pen/xyqEEz?editors=1010

So whenever this is rendered and the translations files are not yet loaded it's not a problem as it will auto-resolve when it does.

I would understand that to reduce complexity to drop Promise support... however I'm not sure if it would be currently possible to provide ONE function/directive that could handle promise results for attribute and node cases?

e.g. would I need something like this?
```js
const msgForAttributes = Directive();
const msgForProperty = Directive();
const msgForNode= Directive();

render(html`
  <my-el label=${msgForAttributes('my-el:label')} .description=${msgForProperty('my-el:label')}>
     <h2>${msgForNode('my-el:heading')}></h2>
  <my-el>
`, target);
```

if so it seem complexity remains => we just push it to our users

=> also creating a promise function is rather "common" so most devs would already know.
Directives on the other hand will need very specific lit-html knowlege... especially if it needs to support node, attribute and properties

=> so overall I am a little torn apart... I understand the need to reduce complexity but I also would like that lit-html becomes easy to use.

**Comment by daKmoR** [comment]
Regarding the `until` directive:
It currently does NOT support attributes (as it assumes lit-html core does it)
https://codepen.io/daKmoR/pen/XxMjGP?editors=1010

I guess we need to put the complexity somwhere. 
So either it's in the core for each "part" (node, attribute, property) or we need to combine them all in the until directive?

if `until` directive supports all of that I guess I could have a `msg` function that "extends" it.
I guess that could work... still all this requires quite in depth knowledge

**Comment by 43081j** [comment]
I do think we should be more tolerant of changes which require "lit specific knowledge".

If we forever try to avoid introducing concepts only we make use of, we'll always be dancing around problems and adding complexity where complexity wouldn't otherwise be needed.

It is ok to expect people to learn how to use directives, especially once we have the docs finished.

**Comment by bgotink** [comment]
@justinfagnani I'm a strong proponent for putting the logic in the `until` directive, as I've already laid out in https://github.com/Polymer/lit-html/issues/544#issuecomment-427660318. Breaking API by removing support for promises in NodePart is a nice benefit in my opinion, because giving people to ways to do the same thing tends to lead to confusion and mistakes.

@daKmoR Multiple until directives would make things overly complex, so I hope we can make it work with the single directive we have now. That would mean your translation API doesn't need to change if you can make `msg` wrap the promise in a call to `until`

**Comment by daKmoR** [comment]
@43081j I may have badly articulated myself. I do agree that having some opinionated solutions will overall improve the situation. (I removed the relevant text from my comments)

@bgotink if that's possible then I'm all for it 👍  I could just replace the return "promise" function with a directive and no additional (user) code changes would be needed - sweet :)

**Comment by LarsDenBakker** [comment]
My vote is for doing promises through directives, I would not call it until though as that implies an in between state which does not always apply. 

It might be better for performance too, because regular renders dont need to do a promise check for each value.

**Comment by justinfagnani** [comment]
If we change the name, I was thinking `wait` or `waitFor` since they're close to `await`.

**Comment by justinfagnani** [comment]
@daKmoR as @43081j points out, you can wrap `until`, so that it's also a directive:

```ts
const msg = (id) => until(_promiseReturningImpl(id));
```

**Comment by bennypowers** [comment]
The notion of 
```js
async render() { ... }
```
was brought up in slack and might be worth some short consideration.

in any event this seems like a positive change.

**Comment by justinfagnani** [comment]
The possibility of writing Promise-returning `asyncRender()` is one of the reasons I like built-in Promise support. `asyncRender` can collect Promise-valued expressions and wait for them while also allowing the incremental rendering. I've been thinking about how to do this with directives, when `render` wouldn't ever see the Promises. It'll require coordination between the directives and render function, I think.

**Comment by 43081j** [comment]
If we rename it, I'd suggest `waitFor` of the two mentioned above. 

An async render method could be nice whichever route we take I think. 

The whole idea of built in promises means having a concept of what should exist by default until they resolve. I like using a directive because consumers have to specify this explicitly. 

I suppose interpolating a promise into a string is a little unusual too. I'd personally expect to see the `toString` of the promise as that's how template literals would do it normally. At least directives make more logical sense as you would expect them to resolve into a string.

**Comment by bennypowers** [comment]
> because consumers have to specify this explicitly.

**Comment by justinfagnani** [comment]
One mark against `waitFor` is that `await` is potentially valid inside a template in an async function and means the _opposite_.

This returns immediately, can render the `<div>` synchronously, and fills in the content with the Promise later:
```ts
const t = (promise) => html`<div>${waitFor(promise)}</div>`;
```

This returns a Promise, can't render the `<div>` synchronously, and waits for the Promise to resolve to return a result:
```ts
const t = async (promise) => html`<div>${await promise}</div>`;
```

**Comment by justinfagnani** [comment]
Huh... `async` is a valid identifier:

```ts
const t = (promise) => html`<div>${async(promise)}</div>`;
```

**Comment by 43081j** [comment]
Sure, we could make the render method async still. It just means you can either incrementally update asynchronously (directives), or render asynchronously (await in render), right?

One is incremental and one is all at once. You could support both

**Comment by 43081j** [comment]
Something like `whenResolved` might work I guess. To avoid confusion. 

it's probably a bad name 😂

**Comment by bennypowers** [comment]
Fwiw until is fine. It's semantic she reads left to right

Until future results, do this.

```
until(promise, default)
```

**Comment by ruphin** [comment]
I am on board with this solution. I've been thinking about how to handle Promises for attributes as well, but it's rather difficult, and as you mentioned there is a lot of undefined behaviour. I'm also not a huge fan of the added complexity to make Promises work in general, since it adds additional work to the hot path of every render action.

I think `wait` and `waitFor` are not good naming schemes, as they signal that the render is blocked somehow. I much prefer the current `until` nomenclature.

**Comment by ruphin** [comment]
Thinking about this more I think this solution will run into the same fundamental problem as Promises. How can a directive know upon resolution of the promise that it has not been superseded by a different render?

Consider the following case

```javascript
template = (thing) => html`
  <div>${thing}</div>
`

render(template(some_promise, 'hello'));
render(template('goodbye'));
```

When `some_promise` resolves, how does `until` know not to override The `'goodbye'` value? It could check the value of the NodePart and compare with the fallback value, but then you still have the same problem with `render(template('hello'))`.

The only way for `until` to know if it is the current rendered value is if `NodePart` sets or clears some property whenever it renders, similar to how rendering Promises is implemented.

**Comment by 43081j** [comment]
The new directive in #555 @ruphin. 

It keeps a weak map of parts to promise state so each time a new promise is passed, the old one does nothing once it resolves. 

There's no problem with default content either as it has to be explicitly specified.

**Comment by ruphin** [comment]
Yes, I understand that if you only ever render to that NodePart through the directive, it can be trivially solved. However, that does mean the NodePart contract becomes more restrictive. (You can render any value, but if you ever render an `until` directive value, you cannot render a non-`until` directive value anymore without breaking)

**Comment by 43081j** [comment]
Do you have a real world use case for doing such a thing?

Sounds like something in need of a refactor to me. Why not use a promise which resolves to your "non-until" value? Why are you doing your switching outside the promise? 

I get your point but it sounds like there are better ways to engineer such a thing.

**Comment by bgotink** [comment]
@justinfagnani I personally don't think the `asyncRender` function requires built-in support for promises. What that function does (dumbed down) is `await` on all promises passed into the `html` template tag before performing a sync `render` operation. Moving promise handling in the sync `render` case into a separate directive is still useful in the case of an `asyncRender` scenario:

```ts
const somePromise = Promise.resolve('value');
const longPromise = new Promise(resolve => setTimeout(resolve, 2000, 'long');

render(html`<div>${somePromise} ${longPromise}</div>`, container);
// <div>[object Promise]</div>
// once microtask queue is flushed: <div>[object Promise]</div>

render(html`<div>${async(somePromise, 'temp')}</div>`, container);
// <div>temp</div>
// once microtask queue is flushed: <div>value</div>

asyncRender(html`<div>${somePromise} ${longPromise}</div>`, container);
// doesn't render
// once microtask queue is flushed: doesn't render
// once 2 seconds pass and microtask queue is flushed: <div>value</div>

// -> This is awesome, I can block the render until promises resolve but I can also tell the renderer to use a temporary value instead
asyncRender(html`<div>${somePromise} ${async(longPromise, 'temp')}</div>`, container);
// doesn't render
// once microtask queue is flushed: <div>value temp</div>
// once 2 seconds pass and microtask queue is flushed: <div>value long</div>
```

**Comment by 43081j** [comment]
@bgotink i dont think your example is quite right..

if we made `render` async (or provided an async version of it), built-in promises still wouldn't be supported as that is a whole different issue/concept...

```js
async render() {
  return html`<div>${somePromise}</div>`;
}
```

This would render the exact same thing as the non-async render call, `[object Promise]`. You interpolated an object, you'd expect the `toString` of it to be rendered.

It is for async work where you don't care about incremental rendering:

```js
async render() {
  const items = await this.getItemsAsync();
  const response = await fetch(url);
  const text = await response.text();
  return html`<div>${text}</div>
    ${items.map((i) => html`<span>${i}</span>`)}`;
}
```

**Comment by ruphin** [comment]
You are thinking of LitElement `render()`, not lit-html `render()`. The examples you show can be supported even with a synchronous render from lit-html.

What I mentioned before is not an issue of supporting certain use cases, it's about the shape of the API. 'The API works fine as long as you use it right' is a bad excuse for having edge cases in your API. A good API produces expected results regardless of what the user does. My point is that the suggested directive adds edge cases that have unexpected results, which adds cost. Either users need to be educated on edge cases which adds documentation complexity and increases the learning curve, or they randomly run into issues and using the library feels like a minefield.

**Comment by 43081j** [comment]
true i am.

i don't see any use in an async lit-html render then, only a lit element one.

I never said the api works fine as long as you use it right.

If you try render an async value then render a non-async value meanwhile (in a second render call), it is perfectly logical to overwrite it once the promise one resolves. Nothing is unexpected about that at all, with basic understanding of promises it is exactly what one would expect.

**Comment by ruphin** [comment]
So you're saying that if you use an API to render a promise, it's expected behaviour that you can't undo or cancel that promise, and you're effectively locked to rendering it once it resolves?

Do you think it's a nice API that the only way to cancel a promise is to render another promise?

**Comment by 43081j** [comment]
Yes I am saying that. It should behave the same way as any async functionality in JS. If you want to prevent it, reject the promise or some such thing.

How do you normally "cancel" a promise? I think you're trying to treat lit differently to a normal async function and shouldn't be. None of this concept is specific to lit, treat it like you do any promise.

**Comment by ruphin** [comment]
It's not about changing the resolution of the promise, but about using the result. I think it's perfectly valid to say that at some point you don't care about the result of a promise anymore. Consider the following scenario:

You want to show a message, with a loading indicator. You implement like this:

```javascript
html`
  <div id="message">
    ${ until(this.message, 'Loading...') }
  </div>
`;
```

Now you want to add localization to this template. You have a `localize()` function that returns a promise, because you don't want your render to be blocked on loading the translations. Your template becomes this:

```javascript
html`
  <div id="message">
    ${ until(this.message, localize('Loading...')) }
  </div>
`;
```

But now things break, because your `localize()` can finish after `this.message`. It seems obvious to me that once you render `this.message` you no longer care about the `localize()` promise, and you don't want to render that anymore.

An other example is a promise on a long-running request. If the user requests some kind of thing and it takes a while to load, and while it's still loading, the user changes the view or requests some other kind of data, you no longer want to render the result of the original request when it resolves.

It's not about promises behaving like in pure JS, it's about API design and user expectation. I expect that if I use an API to render X, and then afterwards render Y, then the result will always be Y. It doesn't matter if X is a promise or not. You are suggesting that I concede all control over what actually happens the moment I decide to render a promise, and what shows up on my screen will be determined by whichever item happens to resolve last. That seems like an unusable API to me.

**Comment by 43081j** [comment]
You're kind of misusing `until`, there, I would say.

The whole point about `async` (until) is to render the result of a promise, but until then, render a default value. It makes no sense for the default value to be _another promise_. This shouldn't even be allowed, the `async` directive exists for the reason of rendering something **until a promise resolves**. where's the sense in using yet another promise? you defeat the whole point of the directive, in that case you'd need a 3rd non-async value to render until both promises resolve.

`async(Promise, primitive)` is the correct signature as far as im concerned.

**Comment by krumware** [comment]
I'm with @43081j here on using the primitive.

Using the extra promise for localization is a great example, but an implementation of that will likely end up just using the same `async(Promise, primitive)`.  So does it make sense to add the complexity to this directive, or implement this logic elsewhere?

**Comment by CaptainCodeman** [comment]
I would much rather lit-html was only concerned with rendering templates and kept as simple as possible with as few directives as possible so it doesn't try to turn into a language itself. Concrete values are converted into DOM nodes / attributes - that should be its sole purpose IMO.

Any handling of Promises, Observables or anything else should be done elsewhere, in higher-order components or some other part of the app (e.g. Redux saga's handling async and producing concrete values to feed the templates).

I get that it will be 'neat' for demos and examples, "look how easy it is to show a loading message, fetch data and then render it!", but IMO it doesn't really add that much that wouldn't be clearer by separating that logic out of the template.

**Comment by ruphin** [comment]
We're getting bogged down into details. The use of `until` in the example was just to illustrate a point. It doesn't matter which deferring mechanism you use, or what the API or intended usage of that mechanism is, there is a simple fundamental problem here.

There are plausible scenarios in which you render a promise of which the result is no longer relevant before it resolves. Whatever system you use, there has to be a way to override that promise, or to stop it from rendering the irrelevant result once it resolves.

As I mentioned before, it is _technically impossible_ to do this, unless either lit-html has some form of builtin support for it, OR you accept an API that breaks if you render any other value after rendering a promise (through whatever system you use, be it an `async` directive, or an `until` directive, or whatever API you want).

--------------------

I think there is a lot of value in lit-html having a deferred rendering system. If you only render static values, that means the entire template needs to re-render whenever any one deferred value resolves. Imagine a page with 10 different values that are all resolved through some higher order logic, redux or whatever. Do you think it's a good idea to just re-render the entire page every time one of these values resolve? I think there clearly is added value in allowing partial deferred rendering, so you only defer the part of the template that has not resolved yet.

**Comment by 43081j** [comment]
`async` is for rendering a non async value until an async value resolves. 

Any variation of that such as with the default being a promise belongs in its own directive I think. 

Simple enough

**Comment by CaptainCodeman** [comment]
> If you only render static values, that means the entire template needs to re-render whenever any one deferred value resolves ... Do you think it's a good idea to just re-render the entire page every time one of these values resolve?

Isn't the point of lit-html that it does this efficiently? (with immutable state update patterns helping that to happen)

**Comment by justinfagnani** [comment]
A bunch of things to tease out here.

First, I think Promises being canceled is fine, but that's a general Promise issue that applies to all of JavaScript. The initial idea there was cancelable Promises, but that stalled (it introduced a new third state to function completion: return, throw and now cancel). The current approach is AbortController/AbortSignal: https://developer.mozilla.org/en-US/docs/Web/API/AbortController as used by `fetch()`. `fetch()` will reject its Promise with an `AbortError` when aborted.

If a Promise passed `until()`/`async()` rejects for any reason, it won't render.

Second, there are some really tricky cases with directives in general that are caused by them being values, and not actually being part of the template. That falls out from our desire to put most features on the expression side of template, and not encode them into strings (binding type being the exception). Directives don't really have any identity or state, so directives that need to keep state, so far, have done so by storing it with the Part. That causes problems with composition, because in a pattern like `async(promise1, async(promise2))`, we can't currently associate state with each directive call separately.

I suspect the vast majority of the time directives are used "statically" (they appear inline with the template) and one-at-a-time, not composed. We have some tests with say, repeat rendering async list items, but directives don't compose that well generally, not without careful handling in the directive implementations.

Third, on async in general, especially in relation to state management - I view lit-html as a low-level, standalone rendering piece that doesn't assume Web Components, or any component library, and doesn't assume state management. At the same time asynchronous values are becoming more and more central to JS development. We can break down the core types of values in JS according to sync/async, and single/multi axes:

|  | Single | Multi |
| --: | :-: | :-: |
| **Sync**  | `T`  | `Iterable<T>` |
| **Async**  | `Promise<T>` | `AsyncIterable<T>` |

I think lit-html should be able to handle all these types of values to some degree - they're that fundamental.

lit-html also follows a pattern where a TemplateResult is just a description of the DOM, that can be rendered by code in another place, at another time, even multiple times. Adding some amount of async handling means that the template definition and the template rendering don't necessarily have to coordinate. Yes, you can remove asynchronicity in templates by re-rendering, but lit-html does not provide a way for a template to tell a renderer to re-render. LitElement does, Redux-like patterns do, but those are not lit-html.

So, I think it's really valuable, even if there are some caveats currently. The caveats are actually with directives in general, and I hope they can be removed in general.

Right now I'm trying to think if there's a way that lit-html can help with managing state for directives so that composition works better. It's possible that directives should work more like `html` and TemplateResults, so that directive arguments are part of the tree of values contained in a TemplateResult - currently only the closure is a value and it's arguments are hidden in the closure. An approach like that might let us pass previous state back to the directive as long as the structure of directives doesn't change between renders - very similar to how we decide to re-render based on the type of TemplateResult rendered to a Part.

**Comment by ruphin** [comment]
Allowing directives to have access to some kind of state helps, but again, it won't solve the particular issue of rendering some value (any value) after rendering some promise-like deferred value (using whatever mechanism).

It's still totally an option to accept that this is simply invalid behaviour and just isn't supported, but I think we should be explicit about making that choice, and it should be explicitly documented.

**Comment by justinfagnani** [comment]
> Allowing directives to have access to some kind of state helps, but again, it won't solve the particular issue of rendering some value (any value) after rendering some promise-like deferred value (using whatever mechanism).

I think it could, by letting a directive query the state object to see if it's still valid. If another value renders there, it could invalidate the state. This is basically a lightweight disconnect signal.

> It's still totally an option to accept that this is simply invalid behaviour and just isn't supported, but I think we should be explicit about making that choice, and it should be explicitly documented.

No matter what, we'll do this. We might even still do that even if we have a design for better stateful directives.

**Comment by LarsDenBakker** [comment]
I'm in favor of offering more hooks to directives about their context / state. The WeakMap bookkeeping takes a lot of knowledge of how directives work, and it's really easy to mess up performance.

**Comment by aadamsx** [comment]
I'm a little confused by this change.  Will someone please do a pros/cons before and after this change (and code would be helpful here)?

**Comment by hyperpress** [comment]
I agree there's been a lot of opinionated drill down as to methods. It
would be helpful if we can see a few before and after code examples on
proposed changes.

**Comment by 43081j** [comment]
Following was suggested (or maybe partially supported already):

```js
const promise = Promise.resolve('foo');
render(html`<div>${promise}</div>`);
// eventually renders <div>foo</div>
```

Following will be required in future to be more explicit about what non-async value to render meanwhile (until promise resolves):

```js
const promise = Promise.resolve('foo');
render(html`<div>${async(promise, 'default')}</div>`);
// renders <div>default</div>
// eventually renders <div>foo</div>
```

**Comment by hyperpress** [comment]
Okay got it. Thanks for the comparison. I'm not seeing anything in the latest proposed change that worries me. I end up doing the most complex logic in redux or similar tools anyway, but using async method would be fine I'd reckon.

**Comment by ruphin** [comment]
@43081j please don't confuse the issue.

First, rendering Promises is currently already supported by lit-html without use of directives. As you can see [here](https://github.com/Polymer/lit-html/blob/ddfd13a4487f2d42056e5a2aa2ab6a629ed07bda/src/lib/parts.ts#L200), NodePart currently has support for rendering Promises, which allow a deferred partial render, or an asynchronous render, if you will. When you render a Promise, lit-html renders nothing, and once the promise resolves, lit-html will render the result of the promise in that location.

The current implementation in lit-html to render Promises is [very inefficient](https://github.com/Polymer/lit-html/issues/451). There is an [outstanding PR](https://github.com/Polymer/lit-html/pull/458) that fixes this performance issue, but it adds some unavoidable complexity to the lit-html core, since fixing Promise rendering essentially boils down to fixing cache invalidation, which is a hard problem in computer science.

Second, there currently already exists a directive that renders a Promise (or async value) with a placeholder (a synchronous value) until it resolves, the [until](https://github.com/Polymer/lit-html/blob/master/src/directives/until.ts) directive. Because lit-html currently supports rendering Promises internally, the implementation of `until` is really trivial.

--------------------------

Now, the proposed change is to move the complexity of rendering Promises out of lit-html core altogether, and into the `until` directive (or a similar directive with another name). 

The advantages of this approach are:

 - It simplifies the core implementation of lit-html. 
 - The new directive would be able to render Promises to both NodeParts _and_ AttributeParts (lit-html currently only supports Promises in NodeParts).

The problems of this approach are:

 - It adds several edge cases that cause unexpected values to render.
 - The solution to this requires directives to have access to some kind of state of the Parts they are rendering to, but they are currently stateless.

@justinfagnani suggests that we may be able to alleviate these problems by tracking some more internal state in Parts, so that directives have more information into the current state of the Part, which solves the above problems.

------------------------

My personal views

The solution suggested by @justinfagnani effectively requires the same type of internal state tracking as the outstanding PR to fix Promises, which means that some complexity in lit-html core is unavoidable. The unfortunate part is that this added complexity adds some performance overhead (I estimate around 1-2%) for every render, but this is _technically unavoidable_ if we want any kind of reliable asynchronous rendering.

However, since tracking this additional state offers added benefits for other directives, _and_ is capable of supporting promises for AttributeParts, I think the suggested solution is clearly a superior implementation over the current PR, because it allows more flexibility for future directives to use.

**Comment by ruphin** [comment]
I never wanted the ability to have promises as default values. I used the example to illustrate a potential problem with this solution.

Let me try again this time with using all the proper syntax as you suggest. If I get the syntax wrong again, please let me know.

Do you think it is a non-issue that the following template can lead to all sorts of weird race conditions when the loggedIn state changes?

```javascript
html`
  <div>
    ${ when(loggedIn, 
         async(asyncContent, 'Loading content'), 
         'Please log in to see your content'
    )}
  </div>
`
```

**Comment by CaptainCodeman** [comment]
> If a Promise passed until()/async() rejects for any reason, it won't render

Presumably just that part not the whole template? No chance to render anything in it's place? Would it be left showing, for instance, a loading message / spinner for a failed fetch?

**Comment by 43081j** [comment]
I think it's an issue for another directive, if any. Not `async` which is being discussed here as a replacement to the built in promise support. 

It is a use case but I think tracking state across async calls is something we should have a dedicated issue for. As it probably affects more than the `async` directive too.

**Comment by ruphin** [comment]
@CaptainCodeman It just won't render the result of the promise. The rest of the render will happen as expected.

**Comment by ruphin** [comment]
@43081j It is not a general directive problem, it is specifically `async` that leads to undefined behaviour when combined with other primitives (any other primitive). Here's the same example, but without other directives this time:

```javascript
html`
  <div>
    ${ loggedIn ?
         async(asyncContent, 'Loading content') :
         'Please log in to see your content'
    )}
  </div>
`
```

Do you still think it's not a problem that this produces unexpected results?

This problem is caused by the implementation of `async`, and it _cannot_ be fixed unless there is specific support for it from lit-html core, for example by adding state tracking, which is what @justinfagnani suggested.

**Comment by 43081j** [comment]
I don't think it's a problem for the async directive to tackle. 

It exists to render a non-async value until an async one resolves. 

It does that. Keep it simple. 

Complex things like state tracking belong in another directive or somewhere else I would say. This directive answers all questions in the issue we are currently in, the original post.

**Comment by 43081j** [comment]
This will simplify the core, not make it more complex.

It doesn't account for many edge cases really either, this discussion got blown up a little. It is an incredibly trivial directive to allow async values, if you want them (entirely optional).
